fs.go

```diff
@@ -65,6 +73,20 @@ type Destroyer interface {
 	Destroy() error
 }
 
+// PinningFS interface for content stores that support the concept of pinnings
+type PinningFS interface {
+	Pin(ctx context.Context, key string, recursive bool) error
+	Unpin(ctx context.Context, key string, recursive bool) error
+}
+
+// CAFS stands for "content-addressed filesystem". Filesystem that implement
+// this interface declare that  all paths to persisted content are reference-by
+// -hash.
+// TODO (b5) - write up a spec test suite for CAFS conformance
+type CAFS interface {
+	IsContentAddressedFilesystem()
```

Comment:
Just thinking out loud here, but it seems like the defining characteristic of CAFS is that the `Put(context, file)` method does not have a path as an input, but instead gets one from the return value. The `file.FullPath` method obscures this a bit, as a local filesystem essentially uses it as another input parameter, whereas a CAFS system needs to ignore it.

But then again, I suppose the same property could also be true for an HTTP based filesystem, wherein a user uploads a file and has no say in where it ends up being available. Maybe a method declaration is insufficient to define exactly what CAFS is, and this is the best we can do.

Reply:
> Just thinking out loud here, but it seems like the defining characteristic of CAFS is that the Put(context, file) method does not have a path as an input

Not quite. If you `Put` a directory, CAFS filesystems rely on paths for their layout. This input:
```
eg
├── a.txt
└── b.txt
```

becomes:
```
$ ipfs add eg -r
added QmbFMke1KXqnYyBBWxB74N4c5SBnJMVAiMNRcGu6x1AwQH eg/a.txt
added QmbFMke1KXqnYyBBWxB74N4c5SBnJMVAiMNRcGu6x1AwQH eg/b.txt
added QmPDmLxkf2T1tErgEy5SnskGcTBRE4BeiyKMHpToihcffw eg

$ ipfs ls QmPDmLxkf2T1tErgEy5SnskGcTBRE4BeiyKMHpToihcffw
QmbFMke1KXqnYyBBWxB74N4c5SBnJMVAiMNRcGu6x1AwQH 0 a.txt
QmbFMke1KXqnYyBBWxB74N4c5SBnJMVAiMNRcGu6x1AwQH 0 b.txt
```

But as you've pointed out, it's hard to define an interface for right now, but we can do it. We can formalize CAFS behaviour by adding tools around the returned path value. The first step is with  tests that assert all  returned values are rooted in a hash and checking the merkle proof.

Doing this at runtime would require doing the same. We can add types / functions for the first "rooted in hash" property, and we already have merkle-proof checks whenever data moves around.